Add SHA-256 and SHA-512 support to fiwalk

jgru opened the pull request:
# Add SHA-256 and SHA-512 support to fiwalk and fix hex digest output

## Description

This pull request adds SHA-256 and SHA-512 hashing support to `fiwalk` and addresses an issue in hex digest formatting (on Apple clang version 15.0.0 (`clang-1500.3.9.4`)).

---

## Changes

### SHA-2 Hash Support
- Added SHA-256 (`-2`) and SHA-512 (`-5`) command-line options.
- Integrated hash generation in `content.cpp` and output via XML, ARFF, and text.
- Updated `-m` and `-z` flags to correctly enable or disable all hash types.

### Access Control Fix
- Made `libcrypto_hasher` methods (`init`, `update`, `finalize`) and the `hash_t` alias `public` to support use in generic hashing wrappers.

### Hex Digest Fixes
- Corrected `std::string` buffer initialization to avoid undefined behavior from `std::string(const char*, size_t)` when using null characters.
- Ensured `hexdigest()` allocates a null-terminated buffer and trims the result correctly to avoid incomplete output.

---

## Included Commits
1. **Add support for SHA256 and SHA512 hashing**
2. **Make libcrypto_hasher methods and type alias public**
3. **Fix hexdigest() buffer allocation to ensure null-termination**
4. **Fix std::string ctor use in hexdigest() to correctly init buffer**

---

## Testing

- Verified output includes SHA-256 and SHA-512 digests when enabled.
- Confirmed hex digest correctness and null-termination.
- Validated correct interactions between hash flags and proper ARFF/XML output.

jgru committed:
Fix std::string ctor use in hexdigest() to correctly init buffer

The previous implementation incorrectly called the std::string(const
char*, size_t) constructor. Because '\0' was interpreted as null, this
results in undefined behavior. This fix ensures the buffer is properly
allocated and zero-initialized for storing the hexadecimal hash digest.

jgru committed:
Fix hexdigest() buffer allocation to ensure null-termination

Allocate buffer with SIZE*2 + 1 to guarantee space for a null terminator
when formatting hex output using snprintf. Trimm final string with
resize() to return only the actual hex digest. This prevents missing or
incomplete output when using .c_str() is used afterward.

jgru committed:
Make libcrypto_hasher methods and type alias public

Expose init, update, finalize methods, and hash_t alias for external
use in generic hash wrappers (i.e., hash_generator__). Without public,
these members are inaccessible due to default private access.

jgru committed:
Add support for SHA256 and SHA512 hashing

This commit
- integrates SHA256 and SHA512 hash generation alongside existing MD5 and SHA-1
- extends the content hashing logic in content.cpp to compute and emit new digests
- updates the command-line options: -2 enables SHA256, -5 enables SHA512
- Updated ARFF and XML output to include SHA256 and SHA512 values
- Adjusted -z and -m options to correctly disable or enable new hash types

simsong commented on tools/fiwalk/src/fiwalk_main.cpp:
```diff
@@ -106,6 +106,8 @@ void usage(fiwalk &o)
     printf("Ways to make this program run slower:\n");
     printf("    -M = Report MD5 for each file (default on)\n");
     printf("    -1 = Report SHA1 for each file (default on)\n");
+    printf("    -2 = Report SHA256 for each file (default off)\n");
+    printf("    -5 = Report SHA512 for each file (default off)\n");
     printf("    -S nnnn = Perform sector hashes every nnnn bytes\n");
```
A cleaner, more extensible approach would have been to modify this to take a name, rather than more single-character options...

jgru replied:
Thank you for the comment, @simsong . Do you want it to be changed to 

```
--sha256       Report SHA256 for each file (default off)
--sha512       Report SHA512 for each file (default off)
```

or would you prefer passing the hash algorithms to use like so  `--hash sha256,sha512`?

simsong replied:
I like the idea of `--hash md5,sha512` ... That's certainly more extensible from a plug-in idea. It probably simplifies the code. But it's a change from what you have here.

jgru replied:
I added support for `--hash alg1,alg2,..` in 915b6099249ad4a4a32a0c909b1ae8886dd12a8c and would appreciate if you have a look. I think it's the best way to go about it. I left the option`-1` and `-M` as they are for backwards compatibility. I you want any further changes, please let me know.

simsong committed:
Merge branch 'develop' into use-sha2-in-fiwalk

simsong committed:
Merge branch 'develop' into use-sha2-in-fiwalk

simsong commented:
BTW — we really appreciate your working on this!

jgru committed:
Fix typos and add spaces

jgru committed:
Use `--hash` option for selecting algorithms

Replaced separate short options (-2 for SHA256, -5 for SHA512) with a new
long option: --hash alg1[,alg2...]
This allows users to specify one or more hash algorithms by name (md5, sha1,
sha256, sha512). Multiple algorithms can be comma-separated.

jgru committed:
Merge branch 'use-sha0-in-fiwalk' of github.com:jgru/sleuthkit into use-sha2-in-fiwalk

jgru commented:
> BTW — we really appreciate your working on this!

Thank you very much, I am happy to help a litte.

simsong reviewed (approved):
Very nicely done.

simsong commented:
Hi @jgru — it looks like your code is failing the mingw compile. Can you view these traces with your account? I'm wondering why they passing for you?

jgru committed:
Use -H instead of --hash to achieve cross-platform compat

Replaces the use of getopt_long() with GETOPT and changes the "--hash"
long option to "-H" in order to be cross-platform compatible.

jgru commented:
> Hi @jgru — it looks like your code is failing the mingw compile. Can you view these traces with your account? I'm wondering why they passing for you?

Hi @simsong , I looked at it. `getopt_long` is not available on Windows, so I changed it to use `-H alg[,alg,..]` (instead of `--hash`) and be cross-platform compatible with minimal changes in [343115d](https://github.com/sleuthkit/sleuthkit/pull/3363/commits/343115db91e0819b5f27e771092fa72c0c93035d). Would you approve another CI run please?
_(For some reason my fork does not have the mingw build in the pipeline.)_

simsong commented:
Running. Thanks for your understanding!

simsong commented:
Great job! Thank you. You are welcome to kick the patch up to 100% code coverage if you want to put on your resume that you made a contribution to an open source project with 100% code coverage of your addition. If not, I can accept this as is.

jgru committed:
Use tabs to indent of case-statements in GETOPT parsing

jgru committed:
Calculate SHA256 and SHA512 in tests to improve code coverage

jgru commented:
> Great job! Thank you. You are welcome to kick the patch up to 100% code coverage if you want to put on your resume that you made a contribution to an open source project with 100% code coverage of your addition. If not, I can accept this as is.

Thank you so much. Aiming to improve code coverage, I added the calculation of SHA256 and SHA512 hashes to `fiwalk_test.cpp` in commit [79c2369](https://github.com/sleuthkit/sleuthkit/pull/3363/commits/79c236951d196f86648959f97a65f244950318ce). 

May I ask for the approval of another CI run so that I can see how much that bumped the code coverage?

jgru committed:
Correct help output to display the short option only

jgru commented:
Hi @simsong, this patch now reaches 81.48% code coverage. Achieving full coverage would require modifying the check_image function in fiwalk_test.cpp to also generate .arff and bodyfiles. Since these outputs have partially conflicting options, additional changes would be needed. I suggest handling that work in a separate PR. What do you think?

simsong commented:
> Hi @simsong, this patch now reaches 81.48% code coverage. Achieving full coverage would require modifying the check_image function in fiwalk_test.cpp to also generate .arff and bodyfiles. Since these outputs have partially conflicting options, additional changes would be needed. I suggest handling that work in a separate PR. What do you think?

I think that you've done a great job and that we can accept this PR!
I have a pet project of getting all of TSK up to 50% code-coverage and had two interns working on it this summer. But we are still a long ways off. 

Thanks again

jgru commented:
> > Hi @simsong, this patch now reaches 81.48% code coverage. Achieving full coverage would require modifying the check_image function in fiwalk_test.cpp to also generate .arff and bodyfiles. Since these outputs have partially conflicting options, additional changes would be needed. I suggest handling that work in a separate PR. What do you think?
> 
> I think that you've done a great job and that we can accept this PR! I have a pet project of getting all of TSK up to 50% code-coverage and had two interns working on it this summer. But we are still a long ways off.
> 
> Thanks again

Awesome! Thank you for taking care of this PR. Looking forward for the merge.
